

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>cfdna.core.cfDNA &mdash; cfdna 0.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/typehints.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/custom.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> cfdna
          

          
          </a>

          
            
            
              <div class="version">
                0.1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api/index.html">The cfDNA API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">cfdna</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>cfdna.core.cfDNA</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for cfdna.core.cfDNA</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">from</span> <span class="nn">ailist</span> <span class="kn">import</span> <span class="n">AIList</span>
<span class="kn">from</span> <span class="nn">NGSfragments</span> <span class="kn">import</span> <span class="n">fragments</span>
<span class="kn">from</span> <span class="nn">NGSfragments.Segment.correction</span> <span class="kn">import</span> <span class="n">correct</span><span class="p">,</span> <span class="n">gaussian_smooth</span>
<span class="kn">from</span> <span class="nn">..</span> <span class="kn">import</span> <span class="n">Plot</span>
<span class="kn">from</span> <span class="nn">..summary</span> <span class="kn">import</span> <span class="n">cfDNA_summary</span>
<span class="kn">from</span> <span class="nn">..multi_summary</span> <span class="kn">import</span> <span class="n">cfDNA_multi_summary</span>


<div class="viewcode-block" id="cfDNA"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.html#cfdna.cfDNA">[docs]</a><span class="k">class</span> <span class="nc">cfDNA</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Wrapper of fragments for cell-free DNA</span>

<span class="sd">	:class:`~cfDNA.cfDNA` stores a fragments object</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sam_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">sam_file</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">min_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
                 <span class="n">paired</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">qcfail</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">mapq_cutoff</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
                 <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize cfDNA object</span>
<span class="sd">                 </span>
<span class="sd">        Params</span>
<span class="sd">        ------</span>
<span class="sd">            sam_fn : str (optional)</span>
<span class="sd">                 Name of the SAM file</span>
<span class="sd">            chrom : str</span>
<span class="sd">                 Name of chromosome</span>
<span class="sd">            sam_file : str?</span>
<span class="sd">                 ?</span>
<span class="sd">            min_size : int</span>
<span class="sd">                 Minimum length of fragment to consider</span>
<span class="sd">            max_size : int</span>
<span class="sd">                 Maximum length of fragment to consider</span>
<span class="sd">            paired : bool</span>
<span class="sd">                 Whether to process as paired reads</span>
<span class="sd">            qcfail : bool</span>
<span class="sd">                 Whether to keep QC failed reads</span>
<span class="sd">            mapq_cutoff : int</span>
<span class="sd">                 Minimum mapping quality allowed</span>
<span class="sd">            proportion : float</span>
<span class="sd">                 Proportion of reads to keep</span>
<span class="sd">            verbose : bool</span>
<span class="sd">                 Print process</span>
<span class="sd">            n_jobs : int</span>
<span class="sd">                 Number of cores to use while reading sam file</span>
<span class="sd">                 </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Read fragments from bam file</span>
        <span class="k">if</span> <span class="n">sam_fn</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">frags</span> <span class="o">=</span> <span class="n">fragments</span><span class="p">(</span><span class="n">sam_fn</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">sam_file</span><span class="p">,</span> <span class="n">min_size</span><span class="p">,</span> <span class="n">max_size</span><span class="p">,</span>
                            <span class="n">paired</span><span class="p">,</span> <span class="n">qcfail</span><span class="p">,</span> <span class="n">mapq_cutoff</span><span class="p">,</span> <span class="n">verbose</span><span class="p">,</span> <span class="n">n_jobs</span><span class="p">,</span> <span class="n">proportion</span><span class="p">)</span>

            <span class="c1"># Calculate length distributions</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">length_dist</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">length_dist</span><span class="p">()</span>

        <span class="c1"># Set attributes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">verbose</span> <span class="o">=</span> <span class="n">verbose</span>


    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">size</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;int: Number of fragments</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">size</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">n_chroms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;int: Number of chromosomes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">n_chroms</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">chroms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;&quot;obj:`list` : List of chromosomes</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">chroms</span>


    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get length from self.frags</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">size</span>

    
<div class="viewcode-block" id="cfDNA.downsample"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.downsample.html#cfdna.cfDNA.downsample">[docs]</a>    <span class="k">def</span> <span class="nf">downsample</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">n_frags</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Randomly downsample fragments</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        proportion : int</span>
<span class="sd">            Proportion of fragments to use</span>
<span class="sd">        n_frags : int</span>
<span class="sd">            Number of fragments to use</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        :class:`~cfDNA.cfDNA`</span>
<span class="sd">            Contains downsamples fragments</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initialize new cfDNA object</span>
        <span class="n">filtered_cfDNA</span> <span class="o">=</span> <span class="n">cfDNA</span><span class="p">()</span>

        <span class="c1"># Filter fragments</span>
        <span class="n">filtered_cfDNA</span><span class="o">.</span><span class="n">frags</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">downsample</span><span class="p">(</span><span class="n">proportion</span><span class="p">,</span> <span class="n">n_frags</span><span class="p">)</span>

        <span class="c1"># Calculate length distributions</span>
        <span class="n">filtered_cfDNA</span><span class="o">.</span><span class="n">length_dist</span> <span class="o">=</span> <span class="n">filtered_cfDNA</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">length_dist</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">filtered_cfDNA</span></div>


<div class="viewcode-block" id="cfDNA.wps"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.wps.html#cfdna.cfDNA.wps">[docs]</a>    <span class="k">def</span> <span class="nf">wps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">protection</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_threads</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Window Protection Score (WPS)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            chrom</span>
<span class="sd">                str</span>
<span class="sd">            protection</span>
<span class="sd">                int</span>
<span class="sd">            min_length</span>
<span class="sd">                int</span>
<span class="sd">            max_length</span>
<span class="sd">                int</span>
<span class="sd">            normalize</span>
<span class="sd">                bool</span>
<span class="sd">            n_threads</span>
<span class="sd">                int</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            wps</span>
<span class="sd">                pandas.Series</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">wps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">wps</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">protection</span><span class="o">=</span><span class="n">protection</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="n">max_length</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="n">normalize</span><span class="p">,</span> <span class="n">n_threads</span><span class="o">=</span><span class="n">n_threads</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">wps</span></div>


<div class="viewcode-block" id="cfDNA.predict_nucleosomes"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.predict_nucleosomes.html#cfdna.cfDNA.predict_nucleosomes">[docs]</a>    <span class="k">def</span> <span class="nf">predict_nucleosomes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">wps</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">merge_distance</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
				            <span class="n">peak_min_length</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">peak_max_length</span><span class="o">=</span><span class="mi">450</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Call peaks from WPS</span>

<span class="sd">		Parameters</span>
<span class="sd">		----------</span>
<span class="sd">			chrom</span>
<span class="sd">				str (chromosomes to call for)</span>

<span class="sd">		Returns</span>
<span class="sd">		-------</span>
<span class="sd">			peaks</span>
<span class="sd">                dict[chrom:AIList] (WPS peaks)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">peaks</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">peak_calling</span><span class="p">(</span><span class="n">wps</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="n">window</span><span class="p">,</span> <span class="n">merge_distance</span><span class="o">=</span><span class="n">merge_distance</span><span class="p">,</span>
				                        <span class="n">peak_min_length</span><span class="o">=</span><span class="n">peak_min_length</span><span class="p">,</span> <span class="n">peak_max_length</span><span class="o">=</span><span class="n">peak_max_length</span><span class="p">,</span>
                                        <span class="n">method</span><span class="o">=</span><span class="s2">&quot;WPS&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">peaks</span></div>

    
<div class="viewcode-block" id="cfDNA.peak_distances"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.peak_distances.html#cfdna.cfDNA.peak_distances">[docs]</a>    <span class="k">def</span> <span class="nf">peak_distances</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">peaks</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_distance</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine distance between peaks</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            peaks</span>
<span class="sd">                dict[chrom:AIList]</span>
<span class="sd">            bin_size</span>
<span class="sd">                int</span>
<span class="sd">            max_distance</span>
<span class="sd">                int</span>
<span class="sd">            smooth</span>
<span class="sd">                bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            peak_bins</span>
<span class="sd">                dict[pandas.Series]</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Calculate distances</span>
        <span class="n">p_dist</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">:</span>
            <span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">peaks</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">intc</span><span class="p">))</span>
            <span class="n">positions</span> <span class="o">=</span> <span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">peaks</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                <span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">peaks</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">end</span>
                <span class="n">positions</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">peaks</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">start</span>
            <span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">positions</span>

        <span class="c1"># Remove those greater than max_distance</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">:</span>
            <span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">=</span> <span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">][</span><span class="n">p_dist</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">max_distance</span><span class="p">]</span>

        <span class="c1"># Bin distances</span>
        <span class="n">peak_bins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">bin_dict_array</span><span class="p">(</span><span class="n">p_dist</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">peak_bins</span><span class="p">:</span>
                <span class="n">peak_bins</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">=</span> <span class="n">gaussian_smooth</span><span class="p">(</span><span class="n">peak_bins</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">peak_bins</span></div>


<div class="viewcode-block" id="cfDNA.fragment_profile"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.fragment_profile.html#cfdna.cfDNA.fragment_profile">[docs]</a>    <span class="k">def</span> <span class="nf">fragment_profile</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bias_correct</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine fragment profile (ratio of large fragments to small fragments)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            chrom</span>
<span class="sd">                str</span>
<span class="sd">            bins_size</span>
<span class="sd">                int</span>
<span class="sd">            genome</span>
<span class="sd">                str</span>
<span class="sd">            bin_bias_h5_fn</span>
<span class="sd">                str</span>
<span class="sd">            bias_correct</span>
<span class="sd">                bool</span>
<span class="sd">            smooth</span>
<span class="sd">                bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            fragment_profile</span>
<span class="sd">                pandas.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Find distributed h5 reference bins</span>
        <span class="k">if</span> <span class="n">genome</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">h5_file</span> <span class="o">=</span> <span class="n">genome</span> <span class="o">+</span> <span class="s2">&quot;_ucsc_bin&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">bin_size</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;.h5&quot;</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reference genome:&quot;</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>
            <span class="n">bin_bias_h5_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">h5_file</span><span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">verbose</span><span class="p">:</span> <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;   Found bin_bias:&quot;</span><span class="p">,</span> <span class="n">h5_file</span><span class="p">)</span>

        <span class="c1"># Calculate coverage of small and large fragments</span>
        <span class="n">small_bin_coverage</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">bin_nhits</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="mi">150</span><span class="p">)</span>
        <span class="n">large_bin_coverage</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">bin_nhits</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="mi">151</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="mi">220</span><span class="p">)</span>

        <span class="c1"># Calculate fragment profile</span>
        <span class="n">fragment_profile</span> <span class="o">=</span> <span class="n">small_bin_coverage</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">fragment_profile</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">small_bin_coverage</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="o">/</span> <span class="n">large_bin_coverage</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="c1"># Correct regions with 0 for both</span>
        <span class="n">fragment_profile</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">fragment_profile</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.0</span>

        <span class="c1"># Correct for bias</span>
        <span class="k">if</span> <span class="n">bias_correct</span><span class="p">:</span>
            <span class="n">fragment_profile</span> <span class="o">=</span> <span class="n">correct</span><span class="p">(</span><span class="n">fragment_profile</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="p">)</span>

        <span class="c1"># Smooth</span>
        <span class="k">if</span> <span class="n">smooth</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">fragment_profile</span><span class="o">.</span><span class="n">seqnames</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
                <span class="n">selected_chrom</span> <span class="o">=</span> <span class="n">fragment_profile</span><span class="o">.</span><span class="n">seqnames</span><span class="o">.</span><span class="n">values</span> <span class="o">==</span> <span class="n">c</span>
                <span class="n">smooth_profile</span> <span class="o">=</span> <span class="n">gaussian_smooth</span><span class="p">(</span><span class="n">fragment_profile</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">selected_chrom</span><span class="p">,</span> <span class="s2">&quot;counts&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
                <span class="n">fragment_profile</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">selected_chrom</span><span class="p">,</span> <span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">smooth_profile</span>

        <span class="k">return</span> <span class="n">fragment_profile</span></div>


<div class="viewcode-block" id="cfDNA.call_cnvs"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.call_cnvs.html#cfdna.cfDNA.call_cnvs">[docs]</a>    <span class="k">def</span> <span class="nf">call_cnvs</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
				  <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">centro_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
				  <span class="n">detail_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outlier_smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">gauss_smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
				  <span class="n">normal</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">],</span> <span class="n">ploidy</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">estimatePloidy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
				  <span class="n">minSegmentBins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">maxCN</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">bcp_cutoff</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Call Copy Number Variants (CNVs)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        chrom : str</span>
<span class="sd">            Name of chromosome</span>
<span class="sd">        bin_size : int</span>
<span class="sd">        min_length : int</span>
<span class="sd">        max_length : int</span>
<span class="sd">        genome : str</span>
<span class="sd">        bin_bias_h5_fn : str</span>
<span class="sd">        genome_fn : str</span>
<span class="sd">        centro_fn : str</span>
<span class="sd">        detail_fn : str</span>
<span class="sd">        outlier_smooth : bool</span>
<span class="sd">        gauss_smooth : bool</span>
<span class="sd">        normal : list</span>
<span class="sd">        ploidy : list</span>
<span class="sd">        estimatePloidy : bool</span>
<span class="sd">        minSegmentBins : int</span>
<span class="sd">        maxCN : int</span>
<span class="sd">        bcp_cutoff : float</span>
<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            cnv : :class:`pandas.DataFrame`</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Call Copy Number Variations</span>
        <span class="n">cnv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">call_cnvs</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="n">max_length</span><span class="p">,</span>
				  <span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="o">=</span><span class="n">bin_bias_h5_fn</span><span class="p">,</span> <span class="n">genome_fn</span><span class="o">=</span><span class="n">genome_fn</span><span class="p">,</span> <span class="n">centro_fn</span><span class="o">=</span><span class="n">centro_fn</span><span class="p">,</span>
				  <span class="n">detail_fn</span><span class="o">=</span><span class="n">detail_fn</span><span class="p">,</span> <span class="n">outlier_smooth</span><span class="o">=</span><span class="n">outlier_smooth</span><span class="p">,</span> <span class="n">gauss_smooth</span><span class="o">=</span><span class="n">gauss_smooth</span><span class="p">,</span>
				  <span class="n">normal</span><span class="o">=</span><span class="n">normal</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">,</span> <span class="n">estimatePloidy</span><span class="o">=</span><span class="n">estimatePloidy</span><span class="p">,</span>
				  <span class="n">minSegmentBins</span><span class="o">=</span><span class="n">minSegmentBins</span><span class="p">,</span> <span class="n">maxCN</span><span class="o">=</span><span class="n">maxCN</span><span class="p">,</span> <span class="n">bcp_cutoff</span><span class="o">=</span><span class="n">bcp_cutoff</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">cnv</span></div>


<div class="viewcode-block" id="cfDNA.window_mean_scores"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.window_mean_scores.html#cfdna.cfDNA.window_mean_scores">[docs]</a>    <span class="k">def</span> <span class="nf">window_mean_scores</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">bed_fn</span><span class="o">=</span><span class="s2">&quot;hg19_TSS.bed&quot;</span><span class="p">,</span> <span class="n">upstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate mean scores for each postion within a window</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            values</span>
<span class="sd">                numpy.ndarray</span>
<span class="sd">            bed_fn</span>
<span class="sd">                str</span>
<span class="sd">            upstream</span>
<span class="sd">                int</span>
<span class="sd">            downstream</span>
<span class="sd">                int</span>
<span class="sd">            stranded</span>
<span class="sd">                bool</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            mean_values</span>
<span class="sd">                numpy.ndarray</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check if bed_fn is in data directory</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">bed_fn</span><span class="p">)):</span>
            <span class="n">bed_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">bed_fn</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="o">~</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">bed_fn</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">FileExistsError</span><span class="p">(</span><span class="s2">&quot;bed_fn does not exist!&quot;</span><span class="p">)</span>
        
        <span class="c1"># Calculate mean scores</span>
        <span class="n">mean_values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">window_mean_scores</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">bed_fn</span><span class="p">,</span> <span class="n">upstream</span><span class="o">=</span><span class="n">upstream</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="n">downstream</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="n">stranded</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">mean_values</span></div>

    
<div class="viewcode-block" id="cfDNA.window_nfr"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.window_nfr.html#cfdna.cfDNA.window_nfr">[docs]</a>    <span class="k">def</span> <span class="nf">window_nfr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">values</span><span class="p">,</span> <span class="n">bed_fn</span><span class="o">=</span><span class="s2">&quot;hg19_TSS.bed&quot;</span><span class="p">,</span> <span class="n">upstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="mi">35</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate Nucleosome Free Region score per window</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            values</span>
<span class="sd">                numpy.ndarray</span>
<span class="sd">            bed_fn</span>
<span class="sd">                str</span>
<span class="sd">            upstream</span>
<span class="sd">                int</span>
<span class="sd">            downstream</span>
<span class="sd">                int</span>
<span class="sd">            stranded</span>
<span class="sd">                bool</span>
<span class="sd">            smooth</span>
<span class="sd">                bool</span>
<span class="sd">            scale</span>
<span class="sd">                bool</span>
<span class="sd">            order</span>
<span class="sd">                int</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            window_values</span>
<span class="sd">                pandas.DataFrame</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Check if bed_fn is in data directory</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">bed_fn</span><span class="p">)):</span>
            <span class="n">bed_fn</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">bed_fn</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="o">~</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">bed_fn</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">FileExistsError</span><span class="p">(</span><span class="s2">&quot;bed_fn does not exist!&quot;</span><span class="p">)</span>
        
        <span class="c1"># Calculate mean scores</span>
        <span class="n">window_values</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">frags</span><span class="o">.</span><span class="n">window_nfr</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">bed_fn</span><span class="p">,</span> <span class="n">upstream</span><span class="o">=</span><span class="n">upstream</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="n">downstream</span><span class="p">,</span>
                                              <span class="n">stranded</span><span class="o">=</span><span class="n">stranded</span><span class="p">,</span> <span class="n">smooth</span><span class="o">=</span><span class="n">smooth</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">order</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">window_values</span></div>

    
<div class="viewcode-block" id="cfDNA.write_summary"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.write_summary.html#cfdna.cfDNA.write_summary">[docs]</a>    <span class="k">def</span> <span class="nf">write_summary</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">summary</span><span class="p">,</span> <span class="n">output_h5</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write summary to h5 file</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            summary</span>
<span class="sd">                cfDNA_summary</span>
<span class="sd">            output_h5</span>
<span class="sd">                str</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Write summary</span>
        <span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output_h5</span><span class="p">)</span></div>


<div class="viewcode-block" id="cfDNA.summarize"><a class="viewcode-back" href="../../../api/cfdna.cfDNA.summarize.html#cfdna.cfDNA.summarize">[docs]</a>    <span class="k">def</span> <span class="nf">summarize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outlier_smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                  <span class="n">gauss_smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">n_threads</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">tss_fn</span><span class="o">=</span><span class="s2">&quot;hg19_TSS.bed&quot;</span><span class="p">,</span> <span class="n">normal</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">],</span>
                  <span class="n">ploidy</span><span class="o">=</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">estimatePloidy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">minSegmentBins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">maxCN</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">bcp_cutoff</span><span class="o">=</span><span class="mf">0.3</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Calculate summary metrics</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">            bin_size</span>
<span class="sd">                int</span>
<span class="sd">            genome</span>
<span class="sd">                str</span>
<span class="sd">            bin_bias_h5_fn</span>
<span class="sd">                str</span>
<span class="sd">            outlier_smooth</span>
<span class="sd">                bool</span>
<span class="sd">            gauss_smooth</span>
<span class="sd">                bool</span>
<span class="sd">            n_threads</span>
<span class="sd">                int</span>
<span class="sd">            tss_fn</span>
<span class="sd">                str</span>
<span class="sd">            normal</span>
<span class="sd">                list</span>
<span class="sd">            ploidy</span>
<span class="sd">                list</span>
<span class="sd">            estimatePloidy</span>
<span class="sd">                bool</span>
<span class="sd">            minSegmentBins</span>
<span class="sd">                int</span>
<span class="sd">            maxCN</span>
<span class="sd">                int</span>
<span class="sd">            bcp_cutoff</span>
<span class="sd">                float</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">            metrics</span>
<span class="sd">                cfDNA_summary</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Initialize metrics</span>
        <span class="n">metrics</span> <span class="o">=</span> <span class="n">cfDNA_summary</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="p">,</span> <span class="n">outlier_smooth</span><span class="p">,</span> <span class="n">gauss_smooth</span><span class="p">,</span> <span class="n">n_threads</span><span class="p">,</span>
                                <span class="n">tss_fn</span><span class="p">,</span> <span class="n">normal</span><span class="o">=</span><span class="n">normal</span><span class="p">,</span> <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy</span><span class="p">,</span> <span class="n">estimatePloidy</span><span class="o">=</span><span class="n">estimatePloidy</span><span class="p">,</span>
				                <span class="n">minSegmentBins</span><span class="o">=</span><span class="n">minSegmentBins</span><span class="p">,</span> <span class="n">maxCN</span><span class="o">=</span><span class="n">maxCN</span><span class="p">,</span> <span class="n">bcp_cutoff</span><span class="o">=</span><span class="n">bcp_cutoff</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">metrics</span></div></div>


<span class="k">def</span> <span class="nf">cfDNA_summarize</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Detect if multiple bams</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">bam</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">:</span>
            <span class="c1"># Define prefix</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">bam</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">prefix</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">bam</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

            <span class="c1"># Create cfDNA object</span>
            <span class="n">cfDNA_object</span> <span class="o">=</span> <span class="n">cfDNA</span><span class="p">(</span><span class="n">sam_fn</span><span class="o">=</span><span class="n">bam</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">proportion</span><span class="p">,</span>
                                 <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_jobs</span><span class="p">,</span>
                                 <span class="n">min_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">min_length</span><span class="p">,</span> <span class="n">max_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">max_length</span><span class="p">)</span>

            <span class="c1"># Check downsampling</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">n_frags</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">cfDNA_object</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">downsample</span><span class="p">(</span><span class="n">n_frags</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_frags</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">proportion</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>
                <span class="n">cfDNA_object</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">downsample</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">proportion</span><span class="p">)</span>

            <span class="c1"># Run CNV for hmm models</span>
            <span class="n">cnv</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">call_cnvs</span><span class="p">(</span><span class="n">chrom</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="mi">1000000</span><span class="p">,</span> <span class="n">min_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_length</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                         <span class="n">genome</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">bin_bias_h5_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">genome_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">centro_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                         <span class="n">detail_fn</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">outlier_smooth</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">gauss_smooth</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                         <span class="n">normal</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">],</span> <span class="n">ploidy</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">estimatePloidy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                         <span class="n">minSegmentBins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">maxCN</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span> <span class="n">bcp_cutoff</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

            <span class="c1"># Summarize metrics</span>
            <span class="n">metrics</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">summarize</span><span class="p">(</span><span class="n">bin_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">n_threads</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_jobs</span><span class="p">)</span>

            <span class="c1"># Relpace HMM metrics</span>
            <span class="n">metrics</span><span class="o">.</span><span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;hmm_loglik&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cnv</span><span class="o">.</span><span class="n">_segments</span><span class="o">.</span><span class="n">hmm_loglik</span>

            <span class="c1"># Plot metrics</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saving summary plot:&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;_summary.pdf&quot;</span><span class="p">)</span>
            <span class="n">Plot</span><span class="o">.</span><span class="n">summary</span><span class="p">(</span><span class="n">metrics</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;_summary.pdf&quot;</span><span class="p">)</span>

            <span class="c1"># Write metrics</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saving summary h5:&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;_summary.h5&quot;</span><span class="p">)</span>
            <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">write_summary</span><span class="p">(</span><span class="n">metrics</span><span class="p">,</span> <span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;_summary.h5&quot;</span><span class="p">)</span>
            
            <span class="c1"># Write seg file</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">segs</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Saving segmentation file:&quot;</span><span class="p">,</span> <span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;.segs&quot;</span><span class="p">)</span>
                <span class="n">seg_fn</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.segs&quot;</span>
                <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">prefix</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">metrics</span><span class="o">.</span><span class="n">write_seg_file</span><span class="p">(</span><span class="n">seg_fn</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Define prefix</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span>

        <span class="c1"># Create cfDNA object</span>
        <span class="n">cfDNA_object</span> <span class="o">=</span> <span class="n">cfDNA</span><span class="p">(</span><span class="n">sam_fn</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bam</span><span class="p">,</span> <span class="n">bin_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">proportion</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">proportion</span><span class="p">,</span>
                            <span class="n">n_frags</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_frags</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_jobs</span><span class="p">)</span>

        <span class="c1"># Check downsampling</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">n_frags</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cfDNA_object</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">downsample</span><span class="p">(</span><span class="n">n_frags</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_frags</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">proportion</span> <span class="o">!=</span> <span class="mf">1.0</span><span class="p">:</span>
            <span class="n">cfDNA_object</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">downsample</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">proportion</span><span class="p">)</span>

        <span class="c1"># Summarize metrics</span>
        <span class="n">metrics</span> <span class="o">=</span> <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">summarize</span><span class="p">(</span><span class="n">bin_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bin_size</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">n_threads</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_jobs</span><span class="p">)</span>

        <span class="c1"># Plot metrics</span>
        <span class="n">Plot</span><span class="o">.</span><span class="n">summary</span><span class="p">(</span><span class="n">metrics</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;_summary.pdf&quot;</span><span class="p">)</span>
        
        <span class="c1"># Write metrics</span>
        <span class="n">cfDNA_object</span><span class="o">.</span><span class="n">write_summary</span><span class="p">(</span><span class="n">metrics</span><span class="p">,</span> <span class="n">prefix</span><span class="o">+</span><span class="s2">&quot;_summary.h5&quot;</span><span class="p">)</span>
        
        <span class="c1"># Write seg file</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">segs</span><span class="p">:</span>
            <span class="n">seg_fn</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;.segs&quot;</span>
            <span class="n">sample</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">prefix</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">metrics</span><span class="o">.</span><span class="n">write_seg_file</span><span class="p">(</span><span class="n">seg_fn</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">cfDNA_merge</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge multiple summary h5 file into one</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Detect directory</span>
    <span class="n">directory_split</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">inputs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">directory_split</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">directory</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">directory</span> <span class="o">=</span> <span class="n">directory_split</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Read multi_summary</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="n">cfDNA_multi_summary</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">inputs</span><span class="p">,</span> <span class="n">directory</span><span class="o">=</span><span class="n">directory</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Write multi_summary</span>
    <span class="n">summary</span><span class="o">.</span><span class="n">write_h5</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">)</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Kyle S. Smith.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>